include/zarray/zreducer_options.hpp

```diff
@@ -0,0 +1,266 @@
+        friend class zreducer;
+
+    private:
+        template<class T>
+        using make_scalar_wrapper_t = zscalar_wrapper<xscalar<T>>;
+
+        template<class EVS>
+        void init_from_tuple(EVS && options);
+
+    public:
+
+
+        zreducer_options(const zreducer_options & ) = default;
+        zreducer_options(zreducer_options && ) = default;
+        zreducer_options & operator= ( const zreducer_options & ) = default;
+
+
+        zreducer_options();
+
+
+        template <class EVS,
+            XTL_REQUIRES(
+                is_reducer_options<std::decay_t<EVS>>
+            )
+        >
+        zreducer_options(EVS && options)
+        :   m_axes(),
+            m_keep_dims(),
+            m_lazy(),
+            m_inital_value()
+        {
+            this->init_from_tuple(std::forward<EVS>(options));
+        }
+
+
+        template <class A, class EVS = DEFAULT_STRATEGY_REDUCERS,
+            XTL_REQUIRES(
+                xtl::negation<is_reducer_options<A>>,
+                is_reducer_options<EVS>,
+                xtl::negation<std::is_same<std::decay_t<A>, zreducer_options>>
+            )
+        >
+        zreducer_options(A&& axes, EVS && options = EVS())
+        :   m_axes(axes.begin(), axes.end()),
+            m_keep_dims(),
+            m_lazy(),
+            m_inital_value()
+        {
+
+            this->init_from_tuple(std::forward<EVS>(options));
+        }
+
+        template <class I, std::size_t N,  class EVS = DEFAULT_STRATEGY_REDUCERS,
+            XTL_REQUIRES(is_reducer_options<EVS>)
+        >
+        zreducer_options( const I(&axes)[N], EVS && options = EVS())
+        :   m_axes(axes, axes+N),
+            m_keep_dims(),
+            m_lazy(),
+            m_inital_value()
+        {
+
+            this->init_from_tuple(std::forward<EVS>(options));
+        }
+
+        template<class T>
+        bool can_get_inital_value() const;
+
+        template<class T>
+        T get_inital_value() const;
+
+        bool has_initial_value() const;
+
+        bool keep_dims() const;
+
+        const auto & axes() const;
+
+        bool is_lazy()const;
+
+
+        template<class T, class F>
+        void visit_reducer_options(const bool force_lazy, F && f)const;
+
+        template<class F>
+        void visit_reducer_options( F && f) const;
+
+    private:
+
+
+        template<class F, class O>
+        void handle_eval_strategie(F && f, const bool force_lazy, O && options) const;
+
+        template<class F, class O>
+        void handle_keep_dim(F && f, O && options) const;
+
+        template<class F, class O>
+        void run(F && f, O && options) const;
+
+        svector<std::size_t> m_axes;
+        bool m_keep_dims;
+        bool m_lazy;
+
+        // it would be nice to replace this with an actual zarray itself
+        // this is not possible since this would lead to cyclic
+        // import / incomplete types.
+        std::shared_ptr<zarray_impl> m_inital_value;
```

Comment:
typo: `m_initial_value`.

I still have a concern with the usage of `shared_ptr` here, it should not be required. What is the cyclic dependency if you include `zarray_zarray.hpp` here?